### PR TITLE
Add visualization of open controlled gates for mpl #3957

### LINKED ISSUE #3957: add visualization of open controlled gates [mpl drawer]
@ewinston says in https://github.com/Qiskit/qiskit-terra/issues/3864:
> In PR #3739, open controlled gates were implemented. With that PR a new keyword argument was added to specify the control state,
> 
> `XGate().control(3, ctrl_state=6)`
> 
> This creates a 4 qubit gate where the first three qubits are controls and the fourth qubit is the target where the x-gate may be applied. A `ctrl_state` may alternatively be expressed as an integer string,
> 
> `XGate().control(3, ctrl_state='110')`
> 
> where it is more apparent which qubits are open and closed. In this case when q[0] is open and q[1] and q[2] are closed the x-gate will be applied to q[4].

As a reference https://github.com/Qiskit/qiskit-terra/pull/3867 implements open controlled gates for the text visualizer like this:
```python
qr = QuantumRegister(4, 'q')
circuit = QuantumCircuit(qr)
circuit.append(U2Gate(pi, -5 * pi / 8).control(3, ctrl_state='100'),
               [qr[0], qr[3], qr[2], qr[1]])
circuit.draw('text')
```
```
                         
q_0: |0>────────o────────
        ┌───────┴───────┐
q_1: |0>┤ U2(pi,-5pi/8) ├
        └───────┬───────┘
q_2: |0>────────■────────
                │        
q_3: |0>────────o────────
```
The current output for `circuit.draw('mpl')` is:
image

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fix #3957

### Details and comments

For open controlled gates, the text drawer shows which control qubits are open or closed. This fix adds the same function for the mpl drawer. In addition the params for the gates are now shown on the gates in the same manner as they would be for a cu3 gate.

Code also had to be added to compute the proper layer_width. The default sections of this code resulted in excessively large layer widths. The new code sets layer_width to 1 if there are 10 or fewer chars in the displayed params, and to 2 if greater than 10. The max is 15 chars. As an example,

```
from qiskit import QuantumRegister, QuantumCircuit
from qiskit.extensions import U3Gate, U2Gate
import numpy as np

pi = np.pi
qr = QuantumRegister(4, 'q')
circuit = QuantumCircuit(qr)
circuit.append(U2Gate(pi/8,pi/8).control(3, ctrl_state=2),
               [qr[0], qr[3], qr[2], qr[1]])
circuit.append(U3Gate(3*pi/4,3*pi/4,3*pi/4).control(3, ctrl_state='101'),
               [qr[0], qr[3], qr[2], qr[1]])
circuit.draw(output='mpl')
```

image

### CODE REVIEW

**Comment by enavarro51** [comment]
@1ucian0, @ajavadia After submitting this, I noticed #4052 and realized the mpl drawer has this problem as well. I have a fix for it, but in the process a question came up. There seems to be an inconsistency in matplotlib.py related to face colors for gates. Should a ControlledGate have the color of the base gate, as in a C3XGate would have the color of an X gate, or should all ControlledGates be the same color? And if they are different colors, should the line colors and the control qubits use those colors as well?

**Comment by enavarro51** [comment]
And along these lines, in the text drawer, the cx and ccx gates display the target cubit as an ordinary X gate. In latex and mpl, it is displayed as a circle with a plus in it. So given the following code, which is the equivalent of a ccx should this use the X gate or the circle-plus?
`circuit.append(XGate().control(2, ctrl_state=3),
               [qr[0], qr[1], qr[2]])`
And if the answer is the circle-plus, then I assume that if ctrl_state != 3, then we would use the X gate since the open control bits would not make this a Toffoli. Thanks.

**Comment by enavarro51** [comment]
Oh, the joys of digging deeply into code. The controlled-Z is handled as a special case and instead of displaying the target qubit as a Z-gate, it displays the target qubit just like a closed control qubit. Any reason for this? Obviously makes it a little hard to distinguish control from target.
`circuit.append(ZGate().control(3, ctrl_state='101'),
               [qr[0], qr[3], qr[2], qr[1]])`
q_0: ─■─
      │ 
q_1: ─■─
      │ 
q_2: ─■─
      │ 
q_3: ─o─

**Comment by ajavadia** [comment]
> There seems to be an inconsistency in matplotlib.py related to face colors for gates.

@enavarro51 for gate colors please see this comment: https://github.com/Qiskit/qiskit-terra/pull/4038#issuecomment-605658480
This PR recently fixed all colors, so I think it should be ok, unless you see something wrong.

> So given the following code, which is the equivalent of a ccx should this use the X gate or the circle-plus?
circuit.append(XGate().control(2, ctrl_state=3), [qr[0], qr[1], qr[2]])
And if the answer is the circle-plus, then I assume that if ctrl_state != 3, then we would use the X gate since the open control bits would not make this a Toffoli.

I think it should use circle-plus regardless of control state. Circle-plus is just a common circuit notation for X. The reason this is not done in the text drawer is that it doesn't render nicely.

> The controlled-Z is handled as a special case and instead of displaying the target qubit as a Z-gate, it displays the target qubit just like a closed control qubit. Any reason for this? Obviously makes it a little hard to distinguish control from target

This is correct. This gate is symmetric, it does not alter the gate if you change control and target. So this is a common way of displaying this fact.

**Comment by enavarro51** [comment]
Ok, got the colors. Also I'll leave the controlled-Z as is, and all Contolled-X will use the circle plus for the target. Thanks again.

**Comment by ajavadia** [comment]
@enavarro51 it looks like cswap gates have inconsistent colors. if you could fix that it would be great. It should be all purple (because it is non-Clifford).

image

**Comment by enavarro51** [comment]
Got it. Let me know if there are any other anomalies since I'm doing a complete roll through the code.

**Comment by ajavadia** [comment]
I also see CU1 and RZZ are displayed in black. These should be purple. Only measurement and reset should be black. 

image

image

**Comment by enavarro51** [comment]
Ok, got those. Is there a reason those 2 gates are done with sidetext instead of in a box like almost all the other gates? Just trying to go for as much consistency as I can.
Also wanted to know if an open cx, cy, etc. is a valid gate, such as,
`circuit.append(XGate().control(1, ctrl_state='0'), [qr[0], qr[1]])`
Thanks.

**Comment by enavarro51** [comment]
Further comments. Updated all the colors to match #4038 list, except for U1, U2, & U3 which were in red and were left that way. rzz and cu1 were the only 2 gates with sidetext, so changed these to the standard - rzz looks like rxx, ryy, etc. and cu1 looks like cu3. Corrected colors on the controlled-swap gate. Fixed some errors in matching the display to the ctrl_state.
Tested all gates listed in qiskit/extensions/standard.

**Comment by ajavadia** [comment]
Thanks. 
U1 u2 u3 we can keep the way it is because those are special IBM basis gates (at least for now). 
For cu1 and rzz, the intention of drawing them like that was that they are symmetric gates, and that’s how they usually render inquantum circuit diagrams. So I think they should be left the way they were.

**Comment by enavarro51** [comment]
Ok. I'll put those 2 back the way they were and recommit.

File: qiskit/visualization/qcstyle.py [hunk 1]
```diff
@@ -96,8 +96,8 @@ def __init__(self):
             'meas': non_gate_color
         }
         self.latexmode = False
-        self.fold = None  # To be removed after 0.10 is released
```
Comment (ajavadia) [inline qiskit/visualization/qcstyle.py:99]: according to this comment, `fold` should just be removed from `qcstyle`. it is now a top-level arg to `circuit.draw()`.

Comment (enavarro51) [reply qiskit/visualization/qcstyle.py:99]: Ok. Here are 3 outputs. Do you want me to remove all references to fold in qcstyle?

image
image
image

**Comment by ajavadia** [comment]
Yes I think fold should be removed from style. 

The colors look but the U3 wodth seems to cause overflows. Can this be fixed?

**Comment by enavarro51** [comment]
Will take care of fold. The overflow in the U3 and others sometimes is due to autoscaling. This is not something I changed. Here's a sample output without autoscaling,
image
I assumed this was expected behavior. Apparently the text is not scaled when the rest of the image is. I can look into this if you'd like. I noticed that the fold for mpl looks pretty clunky compared to the fold for text_drawer. Let me know if you want me to address this or move it to another issue.

**Comment by ajavadia** [comment]
got it. let's open another issue for that then.

I just noticed the control for CX is green (not blue like the rest of the gate). The other controls, e.g. CY or CZ look fine. Can you fix this and then I can approve

image

image

**Comment by enavarro51** [comment]
Ok. I'll take care of these and recommit in the morning.

**Pull request merged**